src/debug.c

```diff
@@ -1210,21 +1210,15 @@ static void* getAndSetMcontextEip(ucontext_t *uc, void *eip) {
     } \
     return old_val; \
 } while(0)
-#if defined(__APPLE__) && !defined(MAC_OS_10_6_DETECTED)
-    /* OSX < 10.6 */
-    #if defined(__x86_64__)
-    GET_SET_RETURN(uc->uc_mcontext->__ss.__rip, eip);
-    #elif defined(__i386__)
-    GET_SET_RETURN(uc->uc_mcontext->__ss.__eip, eip);
-    #else
+#if defined(__APPLE__)
+    /* OSX PowerPC */
+    #if defined(__POWERPC__)
     GET_SET_RETURN(uc->uc_mcontext->__ss.__srr0, eip);
-    #endif
-#elif defined(__APPLE__) && defined(MAC_OS_10_6_DETECTED)
-    /* OSX >= 10.6 */
-    #if defined(_STRUCT_X86_THREAD_STATE64) && !defined(__i386__)
-    GET_SET_RETURN(uc->uc_mcontext->__ss.__rip, eip);
+    /* OSX x86 */
```

Comment:
the history of MAC_OS_10_6_DETECTED is here: https://github.com/redis/redis/commit/a66a4963498d08094fc16c2b0afe6faf8ad512b0

I'm not too fond of the way of fix, we don't have the environment to verify that it's correct, I prefer to do some individual fixes for POWERPC.

Reply:
@sundb How about just moving PowerPC case in front, and then leaving the rest as-is? I am not insisting on code-style at all. Functionally it should be the same, i.e. PowerPC gonna work correctly, the rest will be working as it is now.

Reply:
I.e., add this:
```
#if defined(__APPLE__) && defined(__POWERPC__)
    /* OSX PowerPC */
    GET_SET_RETURN(uc->uc_mcontext->__ss.__srr0, eip);
. . .
```

Reply:
but i saw you removed `MAC_OS_10_6_DETECTED`, does this mean we might compile fail before 10.6?

Reply:
i found that `10.5.8` is the last version of macos for POWERPC, does that mean it must be In `!defined(MAC_OS_10_6_DETECTED)` block?

Reply:
I do not see how having it is justified here. There is no difference between 10.5 and 10.6 (10.4 _is_ different).
I.e. 10.5 will work just like it is (presumably) now, 10.4 will not, but it does not work now anyway. What is the need for `MAC_OS_10_6_DETECTED` check specifically here?

Having said that, I am fine with keeping it, it is inconsequential. Just moving PowerPC case in front of the rest gonna fix the problem. (If you wonder why – 10.6 supports 32-bit `ppc`, but the source ignores that.)

Reply:
Right now if you run the build on 10.6.8 via Rosetta, it will spectacularly break down on this file, since all ARM fallbacks will be used, for PowerPC :)

Reply:
> i found that `10.5.8` is the last version of macos for POWERPC, does that mean it must be In `!defined(MAC_OS_10_6_DETECTED)` block?

This is not the case, strictly speaking. The last official release which natively runs on PowerPC hardware is 10.5.8, but a) some pre-release builds of 10.6 run on PowerPC and b) official releases of 10.6.x support 32-bit ppc via Apple Rosetta (so one can build and use ppc software on an Intel Mac with 10.6 or even in a VM).

Reply:
will this still fail after the fix of `#if defined(__APPLE__) &&MAC_OS_X_VERSION_MAX_ALLOWED >= 1060`?
if we can correctly detect `MAC_OS_10_6_DETECTED` on thePOWERPC, it will not enter the arm block.

Reply:
btw, macos is not actually the target system for our production environment, and we will not have macos release.

Reply:
Macro `#if defined(__APPLE__) && MAC_OS_X_VERSION_MAX_ALLOWED >= 1060` is arch-agnostic. As long as it is used (in its unprefixed form), `MAC_OS_10_6_DETECTED` will be defined on 10.6+.
However, as long as the code hides PowerPC case inside `#ifndef MAC_OS_10_6_DETECTED`, 10.6 build for `ppc` will be broken, since `MAC_OS_10_6_DETECTED` _is_ defined there.

If we want to minimize changes to the ones required, I suggest not modifying any Intel code, but merely checking for PowerPC first (we do not care here if it is 10.5 or 10.6, no reason to). Once PowerPC is checked for, everything further will work as it is now, and no changes are needed in this chunk.

Reply:
it makes sense to me, please also revert other necessary changes.

Reply:
@sundb Sure, please take a look now. I have dropped all non-functional commits and changed the fix for PowerPC to its minimal form.